File: 1026_maximum_difference_between_node_and_ancestor/solution.go
// LeetCode 1026 - Maximum Difference Between Node and Ancestor
// https://leetcode.com/problems/maximum-difference-between-node-and-ancestor/

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func maxAncestorDiff(root *TreeNode) int {
	var dfs func(node *TreeNode, lo, hi int) int
	dfs = func(node *TreeNode, lo, hi int) int {
		if node == nil {
			return hi - lo
		}
		if node.Val < lo {
			lo = node.Val
		}
		if node.Val > hi {
			hi = node.Val
		}
		left := dfs(node.Left, lo, hi)
		right := dfs(node.Right, lo, hi)
		if left > right {
			return left
		}
		return right
	}
	return dfs(root, root.Val, root.Val)
}

File: 1026_maximum_difference_between_node_and_ancestor/solution.js
// LeetCode 1026 - Maximum Difference Between Node and Ancestor
// https://leetcode.com/problems/maximum-difference-between-node-and-ancestor/

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
function TreeNode(val, left, right) {
    this.val = (val === undefined ? 0 : val);
    this.left = (left === undefined ? null : left);
    this.right = (right === undefined ? null : right);
}

/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxAncestorDiff = function(root) {
    const dfs = (node, lo, hi) => {
        if (!node) return hi - lo;
        lo = Math.min(lo, node.val);
        hi = Math.max(hi, node.val);
        return Math.max(dfs(node.left, lo, hi), dfs(node.right, lo, hi));
    };
    return dfs(root, root.val, root.val);
};
